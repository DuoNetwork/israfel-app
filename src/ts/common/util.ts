import { Constants, IUserOrder, Util as CommonUtil } from '@finbook/israfel-common';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from './constants';
import { ICustodianInfo } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = CommonUtil.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'just now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' min ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' hrs ago';
		else return 'long time ago';
	}

	public formatExpiry(isWeek: boolean) {
		return moment(CommonUtil.getExpiryTimestamp(isWeek)).format('YYYY-MM-DD HH:mm');
	}

	public formatTime(time: any) {
		return moment(time).format('MM-DD HH:mm:ss');
	}

	public formatTimeSecond(time: number) {
		return moment(time).format('YYYY-MM-DD HH:mm:ss');
	}

	public formatPriceShort(num: number) {
		if (num < 1) return d3.format('.3f')(num);
		return d3
			.format('.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatStrike(num: number) {
		return d3.format('.4f')(num);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatPercent(num: number) {
		return d3.format('.0%')(num);
	}

	public formatPercentAcc(num: number) {
		return d3.format('.2%')(num);
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatMaturity(maturity: number) {
		return maturity ? moment(maturity).format('YYYY-MM-DD HH:mm') : CST.TH_PERPETUAL;
	}

	public getOrderTitle(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order Placed';
		else if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_FILL) return 'Order Fully Filled';
			else if (order.status === Constants.DB_TERMINATE) return 'Order Expired';
			// order.status === Constants.DB_CONFIRMED
			else return 'Order Cancelled';
		else if (order.status === Constants.DB_MATCHING) return 'Order is Being Filled';
		// if (order.status === Constants.DB_PFILL)
		else return 'Order Partially Filled';
	}

	public getOrderFullDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		let baseDescription = `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}.`;
		if (order.type === Constants.DB_ADD)
			return baseDescription + ` Valid till ${this.formatMaturity(order.expiry)}`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_FILL)
			return baseDescription + ' Fully filled';

		if (order.fill || order.matching) {
			const totalFill =
				order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching);
			baseDescription += ` ${totalFill}(${this.formatPercent(
				totalFill / order.amount
			)}) filled.`;
		}

		if (order.matching && order.type !== Constants.DB_TERMINATE)
			baseDescription += 'Pending settlement';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_CONFIRMED)
			return baseDescription + ' Cancelled by user.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_TERMINATE)
			return baseDescription + ' Expired.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_BALANCE)
			return baseDescription + ` Cancelled due to insufficient balance.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_MATCHING)
			return baseDescription + ` Cancelled due to settlement error.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_RESET)
			return baseDescription + ` Cancelled due to custodian reset.`;

		return baseDescription;
	}

	public getOrderDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		return `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}. Total ${util.formatBalance(
			order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching)
		)} filled.`;
	}

	public getVersionDescription(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order placed.';

		if (order.type === Constants.DB_UPDATE) {
			let description = `Total ${this.formatBalance(order.fill + order.matching)} filled`;
			if (order.matching) description += ', pending settlement';
			return description;
		}

		if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_TERMINATE) return 'Expired';
			else if (order.status === Constants.DB_BALANCE)
				return 'Cancelled due to insufficent balance';
			else if (order.status === Constants.DB_MATCHING)
				return 'Cancelled due to settlement error';
			else if (order.status === Constants.DB_RESET) return 'Cancelled due to custodian reset';
			else if (order.status === Constants.DB_FILL) return 'Fully filled';
			// (order.status === Constants.DB_CONFIRMED)
			else return 'Cancelled by user';
		return 'Invalid order';
	}

	public getMaturityDescription(info?: ICustodianInfo) {
		return `This contract ${
			info && info.code.endsWith('PPT')
				? 'never expires'
				: 'expires in ' +
				(info
					? Math.floor(
							(info.states.maturity - CommonUtil.getUTCNowTimestamp()) / 86400000
						).toFixed(0)
					: 0) +
				' days'
		}.`;
	}

	public convertOrdersToCSV(orders: { [pair: string]: IUserOrder[] }): string {
		const headers = [
			CST.TH_PAIR,
			CST.TH_ORDER_HASH,
			CST.TH_VERSION,
			CST.TH_TIME_UTC,
			CST.TH_SIDE,
			CST.TH_PRICE,
			CST.TH_AMOUNT,
			CST.TH_FILL,
			CST.TH_FEE,
			CST.TH_EXPIRY_UTC,
			CST.TH_STATUS,
			CST.TH_TX_HASH
		];
		let output = '';
		headers.forEach(
			(header, index) =>
				(output += this.wrapCSVString(header) + (index < headers.length - 1 ? ',' : '\n'))
		);
		const ordersByHash: { [orderHash: string]: IUserOrder[] } = {};
		for (const pair in orders)
			orders[pair].forEach(order => {
				if (!ordersByHash[order.orderHash]) ordersByHash[order.orderHash] = [];
				ordersByHash[order.orderHash].push(order);
			});

		const rows: any[][] = [];
		for (const orderHash in ordersByHash) {
			ordersByHash[orderHash].sort((a, b) => a.currentSequence - b.currentSequence);
			ordersByHash[orderHash].forEach((order, index) => {
				const row: any[] = [];
				row.push(order.createdAt);
				row.push(order.pair.replace('|', '-')); // CST.TH_PAIR,
				row.push(order.orderHash); // CST.TH_ID,
				row.push(index + 1); // CST.TH_VERSION,
				row.push(
					moment.utc(order.updatedAt || order.createdAt).format('YYYY-MM-DD HH:mm:ss')
				); // CST.TH_TIME_UTC,
				row.push(order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL); // CST.TH_SIDE,
				row.push(order.price); // CST.TH_PRICE,
				row.push(order.amount); // CST.TH_AMOUNT,
				row.push(order.fill + order.matching); // CST.TH_FILL,
				row.push(order.fee + ' ' + order.feeAsset); // CST.TH_FEE,
				row.push(moment.utc(order.expiry).format('YYYY-MM-DD HH:mm:ss')); // CST.TH_EXPIRY,
				// CST.TH_STATUS,
				if (order.type === Constants.DB_TERMINATE)
					if (order.status === Constants.DB_BALANCE)
						row.push('Cancelled due to insufficent balance');
					else if (order.status === Constants.DB_MATCHING)
						row.push('Cancelled due to settlement error');
					else if (order.status === Constants.DB_RESET)
						row.push('Cancelled due to custodian reset');
					else if (order.status === Constants.DB_FILL) row.push('Fully filled');
					else if (order.status === Constants.DB_CONFIRMED) row.push('Cancelled by user');
					else row.push('Expired');
				else if (order.status === Constants.DB_MATCHING) row.push('Pending settlement');
				else row.push('Open');
				row.push(order.transactionHash || ''); // CST.TH_TX_HASH
				rows.push(row);
			});
		}

		rows.sort((a, b) => a[1].localeCompare(b[1]) || -a[0] + b[0] || -a[3] + b[3]);
		rows.forEach(row =>
			row.forEach((item, index) => {
				if (index)
					output +=
						this.wrapCSVString(item + '') + (index !== row.length - 1 ? ',' : '\n');
			})
		);

		return output;
	}

	public wrapCSVString(input: string): string {
		return ('' + input).indexOf(',') >= 0 ? '"' + input + '"' : input;
	}

	private getEtherScanLink() {
		return `https://${__ENV__ === Constants.DB_LIVE ? '' : 'kovan.'}etherscan.io`;
	}

	public getEtherScanTransactionLink(txHash: string) {
		return `${this.getEtherScanLink()}/tx/${txHash}`;
	}

	public getEtherScanTokenLink(tokenAddr: string, account: string = '') {
		return `${this.getEtherScanLink()}/token/${tokenAddr}${account ? `?a=${account}` : ''}`;
	}

	public getEtherScanAddressLink(account: string) {
		return `${this.getEtherScanLink()}/address/${account}`;
	}

	public fillZero(num: number) {
		return num < 10 ? `0${num}` : `${num}`;
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
