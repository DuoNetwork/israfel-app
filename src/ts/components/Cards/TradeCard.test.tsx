/**
 * @jest-environment jsdom
 */
import { Radio } from 'antd';
import { shallow } from 'enzyme';
// import { mount } from 'enzyme';
import * as React from 'react';
import util from 'ts/common/util';
import { SButton, SInput, SSlider } from './_styled';
import TradeCard from './TradeCard';
const RadioGroup = Radio.Group;

describe('TradeCard Test', () => {
	describe('Test Snapshot', () => {
		const tokenInfo = {
			address: '0x8a3beca74e0e737460bde45a09594a8d7d8c9886',
			code: 'bETH',
			custodian: '0x13016f27945f3f7b39a5daae068d698e34e55491',
			denomination: 0.1,
			feeSchedules: {
				WETH: {
					minimum: 0.1,
					rate: 0
				}
			},
			precisions: { WETH: 0.000005 }
		};
		const tokenBalance = {
			balance: 123,
			allowance: 123,
			address: '0x59E6B3d43F762310626d2905148939973db2BBd3'
		};
		const ethBalance = {
			eth: 123,
			weth: 123,
			allowance: 123
		};
		const orderBook = {
			version: 123,
			pair: 'WETH_ETH',
			bids: [
				{
					price: 123,
					balance: 123,
					count: 123
				}
			],
			asks: [
				{
					price: 123,
					balance: 123,
					count: 123
				}
			]
		};
		const handleClose = jest.fn();
		it('Test Snapshot', () => {
			util.getExpiryTimestamp = jest.fn();
			util.formatFixedNumber = jest.fn();
			util.formatBalance = jest.fn();
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			util.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const addOrder = jest.fn();
			const setUnlimitedTokenAllowance = jest.fn();
			const wrapper = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenInfo={tokenInfo}
					tokenBalance={tokenBalance}
					ethBalance={ethBalance}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			const wrapper1 = shallow(
				<TradeCard
					account={'account'}
					token={'bToken'}
					tokenInfo={tokenInfo}
					tokenBalance={tokenBalance}
					ethBalance={ethBalance}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper1).toMatchSnapshot();
			const wrapper2 = shallow(
				<TradeCard
					account={'account'}
					token={'sToken'}
					tokenInfo={tokenInfo}
					tokenBalance={tokenBalance}
					ethBalance={ethBalance}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper2).toMatchSnapshot();
			const wrapper3 = shallow(
				<TradeCard
					account={'account'}
					token={'dETH'}
					tokenInfo={tokenInfo}
					tokenBalance={tokenBalance}
					ethBalance={ethBalance}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper3).toMatchSnapshot();
			wrapper
				.find(SInput)
				.at(1)
				.simulate('change', { target: { value: '123456' } });
			wrapper
				.find(SInput)
				.at(1)
				.simulate('blur', { target: { value: '123456' } });
			wrapper
				.find(SInput)
				.at(1)
				.simulate('blur', { target: { value: '-123456' } });
			wrapper
				.find(SInput)
				.at(0)
				.simulate('change', { target: { value: '123456' } });
			wrapper
				.find(SInput)
				.at(0)
				.simulate('blur', { target: { value: '123456' } });
			wrapper
				.find(SInput)
				.at(0)
				.simulate('blur', { target: { value: '-123456' } });
			expect(wrapper).toMatchSnapshot();
			wrapper.simulate('keypress', { keyCode: 27 });
			expect(wrapper).toMatchSnapshot();
			wrapper.setState({ approving: false });
			wrapper
				.find(SButton)
				.at(0)
				.simulate('click');
			expect(wrapper).toMatchSnapshot();
		});

		it('Test Snapshot', () => {
			Math.max = jest.fn(() => 1);
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			util.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const addOrder = jest.fn();
			const setUnlimitedTokenAllowance = jest.fn();
			const wrapper = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenBalance={tokenBalance}
					ethBalance={ethBalance}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			wrapper
				.find(SSlider)
				.at(0)
				.simulate('change', { target: { value: '123456' } });
			wrapper
				.find(RadioGroup)
				.at(0)
				.simulate('change', { target: { value: '123456' } });
			expect(wrapper).toMatchSnapshot();
			wrapper
				.find('button')
				.at(0)
				.simulate('click');
			expect(wrapper).toMatchSnapshot();

			wrapper
				.find('button')
				.at(1)
				.simulate('click');
			expect(wrapper).toMatchSnapshot();
			wrapper
				.find(SButton)
				.at(1)
				.simulate('click');

			wrapper.setProps({ token: '' });
			expect(wrapper).toMatchSnapshot();
		});

		it('Approve', () => {
			Math.max = jest.fn(() => 1);
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			util.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const addOrder = jest.fn();
			const setUnlimitedTokenAllowance = jest.fn();
			const wrapper = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenBalance={tokenBalance}
					ethBalance={{
						eth: 123,
						weth: 123,
						allowance: 0
					}}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			wrapper
				.find(SButton)
				.at(0)
				.simulate('click');
		});

		it('getDerivedStateFromProps', () => {
			Math.max = jest.fn(() => 1);
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			util.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const addOrder = jest.fn();
			const setUnlimitedTokenAllowance = jest.fn();
			const wrapper = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenBalance={tokenBalance}
					ethBalance={{ eth: 123, weth: 123, allowance: 0 }}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			wrapper.setProps({ account: 'default ' });
			wrapper.setProps({
				tokenBalance: {
					balance: 123,
					allowance: 123,
					address: '0x59E6B3d43F762310626d2905148939973db2BBd3'
				}
			});
			expect(wrapper).toMatchSnapshot();
		});

		it('limit', () => {
			Math.max = jest.fn(() => 1);
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			util.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const addOrder = jest.fn();
			const setUnlimitedTokenAllowance = jest.fn();
			const wrapper = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenInfo={{
						address: '0x8a3beca74e0e737460bde45a09594a8d7d8c9886',
						code: 'bETH',
						custodian: '0x13016f27945f3f7b39a5daae068d698e34e55491',
						denomination: 0.1,
						feeSchedules: { WETH: { rate: 123, minimum: 123, asset: 'test' } },
						precisions: { WETH: 0.000005 }
					}}
					tokenBalance={tokenBalance}
					ethBalance={{ eth: 0, weth: 0, allowance: 0 }}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			wrapper.setState({ isBid: false });
			expect(wrapper).toMatchSnapshot();

			const wrapper1 = shallow(
				<TradeCard
					account={'account'}
					token={'aETH'}
					tokenInfo={{
						address: '0x8a3beca74e0e737460bde45a09594a8d7d8c9886',
						code: 'bETH',
						custodian: '0x13016f27945f3f7b39a5daae068d698e34e55491',
						denomination: 0.1,
						feeSchedules: {
							WETH: {
								rate: 123,
								minimum: 123
							}
						},
						precisions: { WETH: 0.000005 }
					}}
					tokenBalance={tokenBalance}
					ethBalance={{ eth: 0, weth: 0, allowance: 0 }}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper1).toMatchSnapshot();

			wrapper1.setState({ isBid: false });
			expect(wrapper1).toMatchSnapshot();

			const wrapper2 = shallow(
				<TradeCard
					account={'account'}
					token={'aETH-ETH'}
					tokenInfo={{
						address: '0x8a3beca74e0e737460bde45a09594a8d7d8c9886',
						code: 'bETH',
						custodian: '0x13016f27945f3f7b39a5daae068d698e34e55491',
						denomination: 0.1,
						feeSchedules: { WETH: { rate: 123, minimum: 123, asset: 'test' } },
						precisions: { WETH: 0.000005 }
					}}
					tokenBalance={tokenBalance}
					ethBalance={{ eth: 0, weth: 0, allowance: 0 }}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper2).toMatchSnapshot();
			const wrapper3 = shallow(
				<TradeCard
					account={'account'}
					token={'aETH-ETH'}
					tokenInfo={{
						address: '0x8a3beca74e0e737460bde45a09594a8d7d8c9886',
						code: 'bETH',
						custodian: '0x13016f27945f3f7b39a5daae068d698e34e55491',
						denomination: 0.1,
						feeSchedules: {},
						precisions: { WETH: 0.000005 }
					}}
					tokenBalance={tokenBalance}
					ethBalance={{ eth: 0, weth: 0, allowance: 0 }}
					orderBook={orderBook}
					ethPrice={123}
					navInEth={1}
					navUpdatedAt={1234567890}
					notify={() => ({})}
					interestOrLeverage={1}
					handleClose={handleClose}
					setUnlimitedTokenAllowance={setUnlimitedTokenAllowance}
					addOrder={addOrder}
				/>
			);
			expect(wrapper3).toMatchSnapshot();
		});
	});
});
