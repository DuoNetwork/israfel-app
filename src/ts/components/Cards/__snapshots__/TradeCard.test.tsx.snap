// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradeCard Test Test Snapshot Test Snapshot 1`] = `
<div
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="popup-bg popup-open-bg"
  />
  <Styled(Card)
    className="popup-card popup-open"
    extra={
      <styled.div
        horizontal={true}
        width="40px"
      >
        <img
          className="cardpopup-close"
          src={null}
        />
        <img
          className="cardpopup-close"
          onClick={[MockFunction]}
          src={null}
        />
      </styled.div>
    }
    noBodymargin={true}
    title={
      <styled.div>
        Trade test/WETH
      </styled.div>
    }
    width="360px"
  >
    <div
      className="convert-popup-des"
    >
      <Component
        n={1}
        token="test"
      />
      <div
        className="trade-nav"
      >
        <span
          className="navspan"
        >
          NAV
        </span>
        <span
          className="trade-nav-px"
        >
          1.000000 ETH
        </span>
        <span
          style={
            Object {
              "color": "rgba(64,79,84,.6)",
              "fontSize": 10,
            }
          }
        >
          Last updated: just now
        </span>
      </div>
    </div>
    <styled.div
      horizontal={true}
    >
      <styled.div>
        <div
          className="status-list-wrapper"
        >
          <ul>
            <li
              key="0"
              style={
                Object {
                  "padding": "5px 5px 5px 30px",
                }
              }
            >
              <span
                className="content"
              >
                123.0
              </span>
              <span
                className="title bid-span"
              >
                <b>
                  123.000000
                </b>
              </span>
            </li>
            <li
              key="1"
              style={
                Object {
                  "padding": "5px 5px 5px 30px",
                }
              }
            >
              <span
                className="content"
              >
                -
              </span>
              <span
                className="title bid-span"
              >
                <b>
                  -
                </b>
              </span>
            </li>
            <li
              key="2"
              style={
                Object {
                  "padding": "5px 5px 5px 30px",
                }
              }
            >
              <span
                className="content"
              >
                -
              </span>
              <span
                className="title bid-span"
              >
                <b>
                  -
                </b>
              </span>
            </li>
          </ul>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="status-list-wrapper"
        >
          <ul>
            <li
              key="0"
              style={
                Object {
                  "padding": "5px 30px 5px 5px",
                }
              }
            >
              <span
                className="title ask-span"
              >
                <b>
                  123.000000
                </b>
              </span>
              <span
                className="content"
              >
                123.0
              </span>
            </li>
            <li
              key="1"
              style={
                Object {
                  "padding": "5px 30px 5px 5px",
                }
              }
            >
              <span
                className="title ask-span"
              >
                <b>
                  -
                </b>
              </span>
              <span
                className="content"
              >
                -
              </span>
            </li>
            <li
              key="2"
              style={
                Object {
                  "padding": "5px 30px 5px 5px",
                }
              }
            >
              <span
                className="title ask-span"
              >
                <b>
                  -
                </b>
              </span>
              <span
                className="content"
              >
                -
              </span>
            </li>
          </ul>
        </div>
      </styled.div>
    </styled.div>
    <Spin
      prefixCls="ant-spin"
      size="default"
      spinning={false}
      tip="Pending Order Signature"
      wrapperClassName=""
    >
      <styled.div>
        <styled.div
          horizontal={true}
          padding="10px;"
          width="100%"
        >
          <button
            className="conv-button selected"
            onClick={[Function]}
          >
            BUY
          </button>
          <button
            className="conv-button non-select"
            onClick={[Function]}
          >
            SELL
          </button>
        </styled.div>
        <styled.div
          noMargin={true}
          width="100%"
        >
          <div
            className="status-list-wrapper"
          >
            <ul>
              <li
                style={
                  Object {
                    "padding": "5px 15px",
                  }
                }
              >
                <div
                  className="tabletitle"
                  style={
                    Object {
                      "textAlign": "left",
                      "width": "20%",
                    }
                  }
                >
                  Token
                </div>
                <div
                  className="tabletitle"
                >
                  test
                </div>
                <div
                  className="tabletitle"
                >
                  WETH
                </div>
              </li>
              <li
                style={
                  Object {
                    "padding": "5px 15px",
                  }
                }
              >
                <div
                  className="tabletitle"
                  style={
                    Object {
                      "textAlign": "left",
                      "width": "20%",
                    }
                  }
                >
                  Balance
                </div>
                <div
                  className="tablecontent"
                >
                  123.0
                </div>
                <div
                  className="tablecontent"
                >
                  123.0
                </div>
              </li>
            </ul>
          </div>
        </styled.div>
        <styled.div
          noLiBorder={true}
          noUlBorder={true}
        >
          <div
            className="status-list-wrapper"
          >
            <ul>
              <li
                className="input-line"
                style={
                  Object {
                    "marginBottom": 0,
                    "padding": "0 10px",
                  }
                }
              >
                <span
                  className="input-des"
                >
                  Price
                </span>
                <styled.input
                  min={0}
                  onBlur={[Function]}
                  onChange={[Function]}
                  right={true}
                  step={0.000005}
                  type="number"
                  value=""
                  width="100%"
                />
              </li>
              <li
                className="input-line"
                style={
                  Object {
                    "flexDirection": "row-reverse",
                    "padding": "5px 15px",
                  }
                }
              >
                <span
                  className="des"
                >
                  estimated USD price
                </span>
              </li>
              <li
                className="input-line"
                style={
                  Object {
                    "marginBottom": 0,
                    "padding": "0 10px",
                  }
                }
              >
                <span
                  className="input-des"
                >
                  Amount
                </span>
                <styled.input
                  disabled={true}
                  min={0}
                  onBlur={[Function]}
                  onChange={[Function]}
                  right={true}
                  step={0.1}
                  type="number"
                  value=""
                  width="100%"
                />
              </li>
              <li
                className="input-line"
                style={
                  Object {
                    "padding": "0px 15px",
                  }
                }
              >
                <Styled(Slider)
                  disabled={true}
                  marks={
                    Object {
                      "0": Object {
                        "label": <strong>
                          0%
                        </strong>,
                      },
                      "100": Object {
                        "label": <strong>
                          100%
                        </strong>,
                      },
                      "25": Object {
                        "label": <strong>
                          25%
                        </strong>,
                      },
                      "50": Object {
                        "label": <strong>
                          50%
                        </strong>,
                      },
                      "75": Object {
                        "label": <strong>
                          75%
                        </strong>,
                      },
                    }
                  }
                  onChange={[Function]}
                  step={1}
                  value={0}
                />
              </li>
              <li
                className="input-line"
                style={
                  Object {
                    "padding": "0 15px",
                  }
                }
              >
                <span
                  className="title"
                  style={
                    Object {
                      "width": 40,
                    }
                  }
                >
                  Expiry
                </span>
                <styled.div
                  horizontal={true}
                  rowInv={true}
                  width="46%"
                >
                  <RadioGroup
                    buttonStyle="outline"
                    disabled={false}
                    onChange={[Function]}
                    prefixCls="ant-radio"
                    value={1}
                  >
                    <Radio
                      prefixCls="ant-radio"
                      type="radio"
                      value={1}
                    >
                      Day
                    </Radio>
                    <Radio
                      prefixCls="ant-radio"
                      type="radio"
                      value={2}
                    >
                      Month
                    </Radio>
                  </RadioGroup>
                </styled.div>
                <div
                  className="convert-radio-des"
                >
                  Valid till 1970-01-01 19:00:00
                </div>
              </li>
            </ul>
          </div>
        </styled.div>
      </styled.div>
      <div
        className="convert-popup-des"
      >
        Buy test with WETH
      </div>
      <div
        className="convert-popup-des"
      >
        Pay 0.1 test fee
      </div>
      <styled.div
        horizontal={true}
        padding="10px"
        width="100%"
      >
        <styled.button
          onClick={[Function]}
          width="49%"
        >
          Reset
        </styled.button>
        <styled.button
          disable={true}
          onClick={[Function]}
          style={
            Object {
              "opacity": 0.3,
            }
          }
          width="49%"
        >
          Buy
        </styled.button>
      </styled.div>
    </Spin>
  </Styled(Card)>
</div>
`;
